transport/transport.go

```diff
@@ -222,6 +221,94 @@ type Stream struct {
 	rstError http2.ErrCode
 }
 
+func (s *Stream) setClientStreamReader(consumeConnAndStreamWindows func(uint32), loanSpaceInStreamWindow func(uint32)) {
+	if s.ctx == nil || s.goAway == nil || s.buf == nil {
+		grpclog.Fatalf("Uninitialized stream. s.ctx == nil: %v; s.goAway == nil: %v; s.buf == nil: %v", s.ctx == nil, s.goAway == nil, s.buf == nil)
+	}
+	s.setStreamReader(consumeConnAndStreamWindows, loanSpaceInStreamWindow)
+}
+
+func (s *Stream) setServerStreamReader(consumeConnAndStreamWindows func(uint32), loanSpaceInStreamWindow func(uint32)) {
+	if s.ctx == nil || s.buf == nil {
+		grpclog.Fatalf("Uninitialized stream. s.ctx == nil: %v; s.buf == nil: %v", s.ctx == nil, s.buf == nil)
+	}
+	if s.goAway != nil {
+		grpclog.Fatalf("Unexpected initialization of s.goAway; got %v; want nil", s.goAway)
+	}
+	s.setStreamReader(consumeConnAndStreamWindows, loanSpaceInStreamWindow)
+}
+
+func (s *Stream) setStreamReader(consumeConnAndStreamWindows func(uint32), loanSpaceInStreamWindow func(uint32)) {
+	s.streamReader = &streamWindowSpaceLoaningReader{
+		loanSpaceInStreamWindow: loanSpaceInStreamWindow,
+		wr: &connAndStreamWindowConsumingReader{
+			dec: &recvBufferReader{
+				ctx:    s.ctx,
+				goAway: s.goAway,
+				recv:   s.buf,
+			},
+			consumeConnAndStreamWindows: consumeConnAndStreamWindows,
+		},
+	}
+}
+
+// FullReader makes complete reads into p.
+type FullReader interface {
+	// This Read function blocks until either the len(p) bytes have been read
+	// into p, or until an error occurs.
+	// It's meant to mimick the semantics of the standard libraries io.ReadFull(_, p)
+	// Note that err == nil iff n == len(p).
+	ReadFull(p []byte) (int, error)
+}
+
+// ReadFull reads bytes from the stream and manages flow control as needed.
+// Also note that an error returned by this function indicates an error
+// in the underlying stream, and future calls to Read will continue to
+// return that same error.
+func (s *Stream) ReadFull(p []byte) (n int, err error) {
+	if s.readFullErr != nil {
+		return 0, s.readFullErr
+	}
+	defer func() { s.readFullErr = err }()
```

Comment:
Do you know if the compiler optimizes this?

If not, this would be better:

```go
n, err = io.ReadFull()
s.readFullErr = err
return n, err
```